dbt/adapters/redshift/connections.py

```diff
@@ -323,6 +327,68 @@ def __iam_idc_browser_kwargs(credentials) -> Dict[str, Any]:
 
         return __iam_kwargs(credentials) | idc_kwargs
 
+    def __iam_idc_token_kwargs(credentials) -> Dict[str, Any]:
+        """
+        Expand the logic and flow control here to account for other Identity providers
+        with their specific request configurations. May require new parameters for
+        token_endpoint extraction to distinguish idp's if the url is insufficient.
+
+        We only support token_type=EXT_JWT tokens. token_type=ACCESS_TOKEN has not been
+        tested or is supported but can be added with a presenting use-case.
+        """
+
+        def handle_response(response):
+            """
+            Rate limiting of access_token generation has been called out during refinement
+            and could require special handling, so this method exists to encapsulate that
+            code with an eye towards future handling. All other failures are thrown as
+            normal request failures.
+            """
+            # Handle the 429 rate-limiting case first
+            if response.status_code == 429:
+                raise DbtRuntimeError(
+                    "Rate limit on identity provider's token dispatch has been reached. "
+                    "Consider increasing your identity provider's refresh token rate or "
+                    "lower dbt's maximum concurrent thread count."
+                )
+
+            # Raise for any other non-success status codes (4xx, 5xx)
+            response.raise_for_status()
+
+        logger.debug("Connecting to Redshift with '{credentials.method}' credentials method")
+
+        __validate_required_fields("oauth_token_identity_center", ("token_endpoint",))
+
+        required_keys = {"request_url", "idp_auth_credentials", "request_data"}
+        if required_keys - credentials.token_endpoint.keys():
+            raise FailedToConnectError(
+                "okta requires token_endpoint is provided all three of: \n"
+                "  (1) request_url - the endpoint\n"
+                "  (2) idp_auth_credentials - the base64 encoding of okta client's client_id:client_secret\n"
+                "  (3) request_data field with all necessary parameters in http friendly encoding including refresh_token\n"
+            )
+
+        encoded_idp_client_creds = credentials.token_endpoint["idp_auth_credentials"]
+
+        headers = {
+            "accept": "application/json",
+            "authorization": f"Basic {encoded_idp_client_creds}",
+            "content-type": "application/x-www-form-urlencoded",
+        }
+
+        response = requests.post(
+            credentials.token_endpoint["request_url"],
+            headers=headers,
+            data=credentials.token_endpoint["request_data"],
+        )
+        handle_response(response)
+
+        return __iam_kwargs(credentials) | {
```

Comment:
Minor nit, but I had to look up how the pipe operator works on dicts. Can this just be and `update()` instead?

Reply:
I'd really prefer not to use stateful operations where possible. Expressions are the future 😎

Reply:
🌶️